conventions/cli.md

```diff
@@ -1,79 +1,184 @@
-# Guidelines for `kn` Commands
+## `kn` User Interface Principles
+
+This document describes the conventions that are used for all `kn` commands and options.
+It is normative in the sense that any new feature that introduces or changes user-facing commands or options needs to adhere to the given principles.
+
+Also, the given rules apply to plugins as well.
+Especially any plugin that wants to be promoted to the [knative/client-contrib](https://github.com/knative/client-contrib) plugin repository has to adhere to these rules.
 
-Commands are generally of the form `kn <resource> <verb>`; the resource kind
-forms a command group for all the operations you might want to do with that kind
-of resource.
 
-Commands that directly concern more than one resource kind may be categorized
-with one of the relevant resources, or may get their own top-level verb (eg.
-`connect`).
+### Command Structure
+
+The general format for kn and plugin commands is
+
+```
+kn <noun> [<noun2> ...] <verb> [<id> <id2] [--opt val1 --opt2 --opt3 ...]
+```
 
-Top-level commands concerning the operation of `kn` itself, like `help` and
-`version` are also okay.
+So, commands are generally of the form `kn <noun> <verb>` <sup>[1](#foot-1)</sup>, where [<noun>](#noun) is often the name of a resource (e.g. `service`) but can also refer to other concepts (e.g. `plugin` or `config`).
+This first noun forms a command group for all the operations you might want to do with that kind
+of resource.
+Sometimes there can be deeper hierarchies with multiple nouns (`kn <noun-1> <noun-2> .... <verb>`) when it makes sense to structure complex concepts.
+A good example is `kn source <source-type> <verb>` which is used like in `kn source ping create`.
 
-## Resource
+`kn` commands take only positional arguments that are used as [identifiers](#identifier).
+This identifier is often the name of a resource which identifies the resource uniquely for the current or given namespace.
 
-The <group>.knative.dev Kind, singluar and lowercase. For example, `service` for
-`serving.knative.dev/service` or `trigger` for `eventing.knative.dev/trigger`.
+Top-level commands concerning the operation of `kn` itself, like `help` and `version` are also okay.
 
-## Verb
+### Noun
 
-If the thing the user's doing can be described by the following commands, these
-should be the name of the verb:
+For resource-related commands, the kind itself used as a command in singular and lowercase form. For example, `service` for
+`serving.knative.dev/service` or `trigger` for `eventing.knative.dev/trigger` are the commands for managing these resources respectively
 
-- `describe` prints detailed information about a single resource. It can include
-  status information of related or child resources, too.
+### Verb
 
-- `list` prints summary information about all resources of a type, possibly
-  filtered by parent or label selector.
+If the following subcommands can describe the thing the user's doing following the last noun
```

Comment:
Reads odd. I would use: "If the following subcommands describe an action on the thing or some operation that the user's performing" or similar.

Reply:
Ooops, this was a left over, this line shouldn't be there. Thanks for the heads up !